In [23]:
import torch
import numpy as np

# Bài 1: Tạo Tensor cơ bản
Tạo một Tensor số nguyên kích thước 2 × 3 và hiển thị kết quả. 

In [24]:
import torch
a= torch.tensor([[1,2,3],
                 [4,5,6]],
                 dtype=torch.int32
                 )
print("tensor a: ")
print(a)
print("kich thuoc: ", a.shape)
print("so chieu: ", a.ndim)
print("kieu du lieu: ", a.dtype)


tensor a: 
tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)
kich thuoc:  torch.Size([2, 3])
so chieu:  2
kieu du lieu:  torch.int32


## Nhận xét
 - tensor a có kích thước 2x3
 - tensor a có 2 chiều
 - kiểu dữ liệu của tensor a là torch.int32


# Bài 2: Tạo các Tensor thường dùng
Thực hành tạo Tensor ngẫu nhiên, Tensor toàn số 0, toàn số 1 và ma trận đơn vị. 

In [25]:
a= torch.rand(3,4)
b= torch.randn(3,4)
c= torch.zeros( 2,3)
d= torch.ones(2,3)
e= torch.eye(4)

print(a)
print(b)
print(c)
print(d)
print(e)

tensor([[0.5372, 0.2272, 0.2142, 0.6791],
        [0.1697, 0.1007, 0.6360, 0.4099],
        [0.8184, 0.5733, 0.0222, 0.8317]])
tensor([[ 0.4103, -0.0804,  0.5132,  0.1422],
        [ 0.7472,  0.0765,  0.5923,  1.1471],
        [-1.5265,  2.2720, -0.7869, -1.5278]])
tensor([[0., 0., 0.],
        [0., 0., 0.]])
tensor([[1., 1., 1.],
        [1., 1., 1.]])
tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])


In [26]:

f= torch.full((5,5), 10)
print(f)

tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])


## Nhận xét
- torch.rand() tạo số ngẫu nhiên theo phân phối đều từ 0 đến 1
- torch.randn() tạo số ngẫu nhiên theo phân phối chuẩn
- torch.zeros() tạo tensor toàn số 0
- torch.ones() tạo tensor toàn số 1
- torch.eye() tạo ma trận đơn vị
- torch.full((5,5), 10) tạo tensor 5x5 chứa toàn giá trị 10 

# Bài 3: Kiểm tra thông tin tensor
Sử dụng shape, size(), ndim và dtype để kiểm tra các thuộc tính của Tensor. 

In [27]:
a= torch.zeros(2,3,4)
print("tensor: ")
print(a)
print("shape: ", a.shape)
print("size: ", a.size())
print("number of dimesions: ", a.ndim)
print("data type: ", a.dtype)


tensor: 
tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
shape:  torch.Size([2, 3, 4])
size:  torch.Size([2, 3, 4])
number of dimesions:  3
data type:  torch.float32


## Nhận xét
- torch.size([2,3,4]) cho biết tensor có 3 chiều với kích thước lần lượt là 2,3,4

# Bài 4: Slicing và Indexing
Sử dụng shape, size(), ndim và dtype để kiểm tra các thuộc tính của Tensor. 

In [28]:
a= torch.rand(3,4,5)
print("original tensor: ")
print(a)

print("\ntensor a[1]: ")
print(a[1])

print("\ntensor a[1:, 2:4] :")
print(a[1:, 2:4])

original tensor: 
tensor([[[0.7373, 0.8095, 0.3572, 0.5463, 0.0588],
         [0.4021, 0.3147, 0.3505, 0.8545, 0.7943],
         [0.3077, 0.4812, 0.7564, 0.8049, 0.8682],
         [0.5779, 0.7029, 0.1210, 0.2467, 0.4593]],

        [[0.6451, 0.8441, 0.3577, 0.1799, 0.4264],
         [0.1826, 0.4963, 0.4265, 0.8919, 0.9502],
         [0.0103, 0.6541, 0.2649, 0.6129, 0.0985],
         [0.8098, 0.9458, 0.8896, 0.4252, 0.2029]],

        [[0.2295, 0.9774, 0.8707, 0.7790, 0.7762],
         [0.3024, 0.4042, 0.4439, 0.7324, 0.3075],
         [0.4577, 0.8618, 0.8824, 0.8486, 0.5922],
         [0.9540, 0.3975, 0.7524, 0.7562, 0.0973]]])

tensor a[1]: 
tensor([[0.6451, 0.8441, 0.3577, 0.1799, 0.4264],
        [0.1826, 0.4963, 0.4265, 0.8919, 0.9502],
        [0.0103, 0.6541, 0.2649, 0.6129, 0.0985],
        [0.8098, 0.9458, 0.8896, 0.4252, 0.2029]])

tensor a[1:, 2:4] :
tensor([[[0.0103, 0.6541, 0.2649, 0.6129, 0.0985],
         [0.8098, 0.9458, 0.8896, 0.4252, 0.2029]],

        [[0.4577, 0.861

In [29]:
print("lay 2 phan tu cuoi tren chieu thu 3: ")
print(a[: , : , -2])

lay 2 phan tu cuoi tren chieu thu 3: 
tensor([[0.5463, 0.8545, 0.8049, 0.2467],
        [0.1799, 0.8919, 0.6129, 0.4252],
        [0.7790, 0.7324, 0.8486, 0.7562]])


## Nhận xét
- a có shape là torch.size([3,4,5])
- a[1] có shape là torch.size([4,5])
- a[1:, 2:4] có shape là torch.size([2,2,5])
- a[:, :, -2] là lấy phần tưr ở vị trí số -2 trên chiều thứ 3

# Bài 5: Reshape và Flatten
Thay đổi hình dạng Tensor mà không làm thay đổi số phần tử.

In [30]:
a= torch.rand(3,4)
print("original: ")
print(a)

print("\nFlatten: ")
print(a.ravel())

print("\nReshape to 3x2x2: ")
print(a.reshape(3,2,2))

print("\nReshape thanh kich thuoc 2x6: ")
print(a.reshape(2,6))

print("\nSo phan tu ban dau: ", a.numel())
print("\nSo phan tu sau khi reshape 3x2x2: ", a.reshape(3,2,2).numel())
print("\nSo phan tu sau khi reshape 2x6: ", a.reshape(2,6).numel())

original: 
tensor([[0.7702, 0.0037, 0.2384, 0.6442],
        [0.5493, 0.2081, 0.8957, 0.9470],
        [0.5922, 0.8333, 0.1870, 0.7403]])

Flatten: 
tensor([0.7702, 0.0037, 0.2384, 0.6442, 0.5493, 0.2081, 0.8957, 0.9470, 0.5922,
        0.8333, 0.1870, 0.7403])

Reshape to 3x2x2: 
tensor([[[0.7702, 0.0037],
         [0.2384, 0.6442]],

        [[0.5493, 0.2081],
         [0.8957, 0.9470]],

        [[0.5922, 0.8333],
         [0.1870, 0.7403]]])

Reshape thanh kich thuoc 2x6: 
tensor([[0.7702, 0.0037, 0.2384, 0.6442, 0.5493, 0.2081],
        [0.8957, 0.9470, 0.5922, 0.8333, 0.1870, 0.7403]])

So phan tu ban dau:  12

So phan tu sau khi reshape 3x2x2:  12

So phan tu sau khi reshape 2x6:  12


## Nhận xét:
- tensor ban đầu 3x4 là 12
- sau khi reshape 3x2x2 là 12
- sau khi reshape 2x6 là 12
- vậy reshape chỉ thay đổi hình dạng của tensor chứ không thay đổi tổng số phần tử

# Bài 6: Phép toán và nhân ma trận
Thực hiện các phép toán phần tử và phép nhân ma trận bằng toán tử @. 

In [31]:
a= torch.randn(2,3)
b= torch.randn(2,3)
print("a: ")
print(a)
print("\nb: ")
print(b)
print("\na + b: ")
print(a+b)
print("\na / b: ")
print(a/b)
print("\na^2: ")
print(a**2)
print("\nmatrix multiplication( a@b.T): ")
print(a @ b.T)

print("\nshape cua a@b.T : ")
print((a@b.T).shape)

a: 
tensor([[ 0.5155, -0.2648,  0.1462],
        [ 1.3076,  0.5090,  0.6180]])

b: 
tensor([[-0.8088,  0.1774, -1.2006],
        [-1.5099,  1.5267,  0.3703]])

a + b: 
tensor([[-0.2933, -0.0873, -1.0545],
        [-0.2023,  2.0356,  0.9883]])

a / b: 
tensor([[-0.6374, -1.4921, -0.1217],
        [-0.8660,  0.3334,  1.6690]])

a^2: 
tensor([[0.2658, 0.0701, 0.0214],
        [1.7099, 0.2591, 0.3820]])

matrix multiplication( a@b.T): 
tensor([[-0.6394, -1.1285],
        [-1.7093, -0.9686]])

shape cua a@b.T : 
torch.Size([2, 2])


## Nhận xét:
- phép toán a*b là nhân từng phần tử tương ứng
- phép toán a@b.T là nhân ma trận 
- với a có kích thước 2x3 còn b.T có kích thước 3x2, nên shape của kết quả a@b.T là 2x2


# Bài 7: Các hàm thống kê trên Tensor
Tính trung bình, độ lệch chuẩn và tổng tích lũy theo từng cột. 

In [32]:
a= torch.randn(3,4)
print(a)
print("\nMean: ")
print(torch.mean(a, dim=0))

print("\nStandard deviation: ")
print(torch.std(a, dim=0))

print("\nCumulative sum: ")
print(torch.cumsum(a, dim=0))


print("\nMean: ")
print(torch.mean(a, dim=1))

print("\nStandard deviation: ")
print(torch.std(a, dim=1))

print("\nCumlative sum: ")
print(torch.cumsum(a, dim=1))

tensor([[ 0.6063,  0.8588,  0.3820,  1.1036],
        [-0.7957, -0.2872,  2.3275, -1.1762],
        [ 0.2368,  0.1024,  1.4040,  0.7120]])

Mean: 
tensor([0.0158, 0.2247, 1.3711, 0.2132])

Standard deviation: 
tensor([0.7266, 0.5827, 0.9732, 1.2191])

Cumulative sum: 
tensor([[ 0.6063,  0.8588,  0.3820,  1.1036],
        [-0.1894,  0.5716,  2.7095, -0.0726],
        [ 0.0474,  0.6740,  4.1134,  0.6395]])

Mean: 
tensor([0.7377, 0.0171, 0.6138])

Standard deviation: 
tensor([0.3122, 1.5827, 0.5881])

Cumlative sum: 
tensor([[ 0.6063,  1.4651,  1.8471,  2.9507],
        [-0.7957, -1.0829,  1.2446,  0.0684],
        [ 0.2368,  0.3392,  1.7432,  2.4552]])


## Nhận xét:
- dim=0 thực hiện theo chiều đầu tiên nên kết quả thống kê theo từng cột
- dim=1 thực hiện theo chiều thứ 2 nên kết quả thống kê theo từng hàng
- khi đổi dim=0 sang dim=1, phép tính thực hiện theo chiều khác nhau nên kết quả chuyển từ thống kê theo cột sang thống kê theo hàng

# Bài 8: Autograd cơ bản
Tạo biến có requires_grad=True và dùng backward() để tính đạo hàm tự động. 

In [33]:
x= torch.tensor(
    3.6,
    requires_grad=True
)
y= x*x
y.backward()

print("x= ", x)
print("y= ", y)
print("x.grad= ", x.grad)

x=  tensor(3.6000, requires_grad=True)
y=  tensor(12.9600, grad_fn=<MulBackward0>)
x.grad=  tensor(7.2000)


## kiểm tra gradient
-  ta có: y = x²
- đạo hàm y'= 2x
- tại x=3.6 thì y'= 2 x 3.6 = 7.2
- kết quả bằng với x.grad 

## Đổi hàm  thành y = 3*x**2 + 2*x + 1 và kiểm tra gradient.

In [34]:
x= torch.tensor(
    3.6,
    requires_grad=True
)
y= 3*x**2+2*x+1
y.backward()

print("x= ", x)
print("y= ", y)
print("x.grad= ", x.grad)


x=  tensor(3.6000, requires_grad=True)
y=  tensor(47.0800, grad_fn=<AddBackward0>)
x.grad=  tensor(23.6000)


## Kiểm tra gradient
- ta có: y = 3x²+2x+1
- đạo hàm: y'= 6x+2
- tại x=3.6 thì y'= 6 x 3.6 + 2= 23.6

# Bài 9: Gradient Descent tìm hệ số đa thức
- Sinh dữ liệu từ đa thức y = x² + 2x + 3, sau đó dùng NAdam và autograd để tìm lại ba hệ số.

In [35]:
polynomial= np.poly1d([1,2,3]) #đa thức thật
N=20 # số lượng dữ liệu

X= np.random.randn(N,1)*5 #sinh dữ liệu X
Y= polynomial(X) #tính Y theo đa thức

XX= np.hstack([  # tạo ma trận đặc trưng [x^2, x, 1]
    X*X,
    X,
    np.ones_like(X)
])

w= torch.randn(3,1, requires_grad=True) #3 hệ số cần tìm

#chuyển dữ liệu sang tensor
x= torch.tensor(XX, dtype=torch.float32)
y= torch.tensor(Y, dtype=torch.float32)

optimizer=torch.optim.NAdam([w], lr=0.01)

print("initial coeficients: ")
print(w)

# huấn luyện
for epoch in range (1000):
    optimizer.zero_grad() # xoá gradient của lần lặp trước
    y_pred= x@w
    mse= torch.mean(torch.square(y-y_pred))
    mse.backward() # tính gradient của hàm mất mát mse đối với các hệ số
    optimizer.step() #dùng gradient để cập nhật các hệ số, làm cho sai số nhỏ hơn

print("learned coeficients: ")
print(w)

print("\nExpected coeficients: ")
print("[1,2,3]")



initial coeficients: 
tensor([[1.6893],
        [0.2710],
        [0.1123]], requires_grad=True)
learned coeficients: 
tensor([[1.0092],
        [2.0095],
        [2.0327]], requires_grad=True)

Expected coeficients: 
[1,2,3]


In [36]:
print("\nSo sanh he so:")

print("He so tim duoc:")
print(w.detach().numpy())

print("\nHe so mong doi:")
print("[1, 2, 3]")


So sanh he so:
He so tim duoc:
[[1.0091531]
 [2.0094535]
 [2.0326614]]

He so mong doi:
[1, 2, 3]


## So sánh 3 hệ số học được với [1,2,3]
- nhận được 3 hệ số gần với [1,2,3]
## Giải thích vai trò của  optimizer.zero_grad(), mse.backward() và optimizer.step(). 
- optimizer.zero_grad():  xoá gradient của lần lặp trước
- mse.backward(): tính gradient của hàm mất mát mse đối với các hệ số
- optimizer.step(): dùng gradient để cập nhật các hệ số nhằm làm cho sai số nhỏ hơn

# Bài 10: Giải hệ phương trình bằng Autograd 
- Tối thiểu hóa tổng sai số bình phương để tìm A, B, C, D thỏa bốn phương trình. 

In [37]:
import random
A= torch.tensor(random.random(), requires_grad=True)
B= torch.tensor(random.random(), requires_grad=True)
C= torch.tensor(random.random(), requires_grad=True)
D= torch.tensor(random.random(), requires_grad=True)

EPOCHS =2000
optimizer= torch.optim.NAdam([A,B,C,D], lr=0.01)

for _ in range(EPOCHS):
   #sai số của phương trình
   y1= A+B-9
   y2= C-D-1
   y3= A+C-8
   y4= B-D-2

   #tổng bình phương sai số
   sqerr= y1 * y1 + y2 * y2 + y3 * y3 + y4 * y4

   optimizer.zero_grad()
   sqerr.backward()
   optimizer.step()

print("A= ", A)
print("B= ", B)
print("C= ", C)
print("D= ", D)



A=  tensor(4.3531, requires_grad=True)
B=  tensor(4.6469, requires_grad=True)
C=  tensor(3.6469, requires_grad=True)
D=  tensor(2.6469, requires_grad=True)


In [38]:
print("\nKiem tra he phuong trinh: ")
print("A + B= ", (A+B).item())
print("C - D= ", (C-D).item())
print("A + C= ", (A+C).item())
print("B - D= ", (B-D).item())


Kiem tra he phuong trinh: 
A + B=  8.999953269958496
C - D=  1.000016689300537
A + C=  7.999982833862305
B - D=  1.9999868869781494


## Nhận xét:
- hệ phương trình có nhiều nghiệm vì có 4 ẩn nhưng các phương trình không độc lập để xác định duy nhất cả 4 ẩn. có thể chọn biến tự do, các biến còn lại được xác định theo biến đó

# Bài 11:  Viết chương trình PyTorch để thực hiện các yêu cầu sau: 
• Tạo Tensor X có 24 phần tử và reshape lần lượt thành 4 × 6, 2 × 12 và 2 × 3 × 4. 
• Tạo hai Tensor A, B kích thước 3 × 4; tính A+B, A*B, A/B, trung bình theo cột và độ lệch chuẩn theo cột. 
• Cho y = 2x² + 5x + 3. Dùng autograd tính dy/dx tại x = 2 và kiểm tra bằng phép tính thủ công. 
• Thay đa thức trong Bài 9 bằng y = 2x² - 3x + 5; huấn luyện và cho biết các hệ số tìm được. 

## Tensor 24 phần tử


In [39]:
X= torch.arange(24)
print("X: ")
print(X)

print("\nReshape 4x6 :\n", X.reshape(4,6))
print("\nReshape 2x12 :\n", X.reshape(2,12))
print("\nReshape 2x3x4:\n ", X.reshape(2, 3, 4) )


X: 
tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23])

Reshape 4x6 :
 tensor([[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23]])

Reshape 2x12 :
 tensor([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]])

Reshape 2x3x4:
  tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])


## Tensor A và B

In [40]:
A= torch.randn(3,4)
B= torch.randn(3,4)

print("A: ")
print(A)
print("\nB: ")
print(B)

print("\nA + B:\n ", (A + B))
print("\nA * B:\n ", (A * B))
print("\nA / B:\n ", (A / B))

print("\nTrung binh theo cot: ")
print(torch.mean(A, dim=0))

print("\nDo lech chuan theo cot: ")
print(torch.std(A, dim=0))


A: 
tensor([[-0.0345,  0.3144, -1.1486,  0.0918],
        [-1.0305,  0.0879, -0.5532, -0.0660],
        [-0.0806, -1.8474,  0.5135, -0.6347]])

B: 
tensor([[ 0.6905,  0.4900, -0.3921, -0.7915],
        [ 1.1601,  0.2825, -1.5360, -2.5702],
        [ 1.2264,  0.8954, -0.5798,  2.1934]])

A + B:
  tensor([[ 0.6560,  0.8044, -1.5407, -0.6998],
        [ 0.1297,  0.3704, -2.0892, -2.6362],
        [ 1.1457, -0.9520, -0.0662,  1.5587]])

A * B:
  tensor([[-0.0238,  0.1541,  0.4504, -0.0726],
        [-1.1954,  0.0248,  0.8496,  0.1696],
        [-0.0989, -1.6542, -0.2977, -1.3922]])

A / B:
  tensor([[-0.0499,  0.6417,  2.9291, -0.1159],
        [-0.8882,  0.3113,  0.3601,  0.0257],
        [-0.0657, -2.0633, -0.8858, -0.2894]])

Trung binh theo cot: 
tensor([-0.3818, -0.4817, -0.3961, -0.2030])

Do lech chuan theo cot: 
tensor([0.5622, 1.1882, 0.8421, 0.3821])


## Dùng Autogrand

In [41]:
x= torch.tensor(2.0, requires_grad=True)

y= 2* x**2 + 5 * x + 3
y.backward()

print("x= ", x.item())
print("y= ", y.item())
print("y'= ", x.grad.item())


x=  2.0
y=  21.0
y'=  13.0


### Kiểm tra bằng thủ công: y= 2x² + 5x + 3 tại x= 2
- y'= 4x + 5
- tại x=2 thì y'= 4*2 +5 = 13


## Học đa thức mới

In [42]:
polynomial = np.poly1d([2, -3, 5])
N=20

# Sinh dữ liệu
X= np.random.randn(N,1)*5

# tính y
Y=polynomial(X)

# tạo ma trận 
XX= np.hstack([
    X * X,
    X,
    np.ones_like(X)
])

# Khởi tạo 3 hệ số
ư= torch.randn(
    3,1,
    requires_grad=True
)

# chuyển sang tensor
x_train= torch.tensor(
    XX,
    dtype=torch.float32
)

y_train= torch.tensor(
    Y,
    dtype=torch.float32
)

# NAdam
optimizer = torch.optim.NAdam(
    [w],
    lr=0.01
)

print("he so ban dau: ")
print(w)

# huan luyen
for epoch in range(1000):
    optimizer.zero_grad()
    y_pred= x_train @ w
    mse = torch.mean(
        torch.square(y_train - y_pred)
    )
    mse.backward()
    optimizer.step()

print("\nHe so hoc duoc: ")
print(w)

print("\nHe so mong muon: ")
print("[2, -3, 5]")

he so ban dau: 
tensor([[1.0092],
        [2.0095],
        [2.0327]], requires_grad=True)

He so hoc duoc: 
tensor([[ 2.0581],
        [-2.1403],
        [ 3.2500]], requires_grad=True)

He so mong muon: 
[2, -3, 5]
